From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 3 Aug 2025 22:29:34 +0800
Subject: [PATCH] Skip BlockPhysicsEvent if no listeners

BlockPhysicsEvent creates new CraftBlock instance and clones CraftBlockData when constructed it.
It's a defensive programming stratgy and expensive if there are many neighbor updates.
However, there is no plugin listening to this event sometimes, so that just add a check to not
call it if no listeners listening to it.

diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 87512434ccd8d46d8db718228c790465c64b2c20..7afbff76005062b4d0dee94c6d5fdb0ea6224de9 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -49,6 +49,7 @@ public interface NeighborUpdater {
     static void executeUpdate(Level level, BlockState state, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston, BlockPos sourcePos) {
         // Paper end - Add source block to BlockPhysicsEvent
         try {
+            if (((net.minecraft.server.level.ServerLevel) level).hasPhysicsEvent) { // Leaf - Skip BlockPhysicsEvent if no listeners
             // CraftBukkit start
             org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(state), org.bukkit.craftbukkit.block.CraftBlock.at(level, sourcePos)); // Paper - Add source block to BlockPhysicsEvent
             level.getCraftServer().getPluginManager().callEvent(event);
@@ -57,6 +58,7 @@ public interface NeighborUpdater {
                 return;
             }
             // CraftBukkit end
+            } // Leaf - Skip BlockPhysicsEvent if no listeners
             state.handleNeighborChanged(level, pos, neighborBlock, orientation, movedByPiston);
             // Spigot start
         } catch (StackOverflowError ex) {
